### ISSUE TITLE
Improve configuration code

### ISSUE THREAD

**Issue by bartekn** [stellar/go#810]
Adding new config value to Horizon requires developers to add new code in 3 different places. Ex. `ENABLE_ASSET_STATS` feature:
* Bind env variable to param switch: https://github.com/stellar/go/blob/b3cacf3189c40fc2ff68bd45fadd5716b9ccc83f/services/horizon/main.go#L54
* Add new flag: https://github.com/stellar/go/blob/b3cacf3189c40fc2ff68bd45fadd5716b9ccc83f/services/horizon/main.go#L199
* Add to `Config` struct: https://github.com/stellar/go/blob/b3cacf3189c40fc2ff68bd45fadd5716b9ccc83f/services/horizon/main.go#L311

It's very easy to make a mistake or forget to add new flag to one of the places above. Ideally there should be a single function that registers new flag in all required places.

**Comment by ire-and-curses** [comment]
I'm thinking of an array of `config` structs. Each struct corresponds to a command line arg configuration. A configuration pseudocode looks like this:

```
type flagType func()
const (
    stringFlag flagType = rootCmd.PersistentFlags().String
    intFlag flagType = rootCmd.PersistentFlags().Int
    boolFlag flagType = rootCmd.PersistentFlags().Bool
)
var anything interface{}

type Config struct {
    name string
    envVar string
    flagType flagType
    default anything
    description string
    configKey string
    configValue func
}
```

For example, `enable-asset-stats` would be configured like this:
```
    enableAssetStats := Config{
    name: "enable-asset-stats"
    envVar: "ENABLE_ASSET_STATS"
    flagType: boolFlag
    default: false
    description: "enables asset stats..."
    configKey: "EnableAssetStats"
    configValue: viper.GetBool("enable-asset-stats")
}
```

Given `name`, we can derive `envVar` and `configValue` automatically for most variables, so helpers  can be provided to default these for 90% of cases. We have to allow them to be overriden though, since they are different in a couple of cases (`db-url` -> `DATABASE_URL`).

`configValue` can be a custom anonymous function (e.g. for `sse-update-frequency`). We will pass the `name` to whatever function is provided.

Then a parser evaluates this configuration, and updates `viper.BindEnv`, `rootCmd.PersistentFlags`, and the `horizon.Config` object.

This is just a rough plan. Your thoughts, suggestions and criticisms are welcome!

**Comment by bartekn** [comment]
Looks good! Some suggestions:
* Can `configValue` be automatically fetched given there are `flagType` and `name` fields?
* For `configKey` - I guess we will need to use reflection. Maybe a simple pointer to the value would be better? Extra advantage: protection from typos.
* We could add `required` field and if zero-value is passed we can autogenerate error message like:
  ```
  Invalid config: network-passphrase is blank.  Please specify --network-passphrase on the command line or set the NETWORK_PASSPHRASE environment variable.
  ```

**Comment by ire-and-curses** [comment]
Great suggestions! I didn't quite understand what you meant for number 2: `Maybe a simple pointer to the value would be better? Extra advantage: protection from typos.` Are you saying we should change Config from a `struct` to something else?

**Comment by bartekn** [comment]
I was thinking about changing `configKey` from `string` to `interface{}` (that is actually pointer to a value in `Config` struct). But after writing a [sample code](https://play.golang.org/p/CytDH4DiMfW) I realized that there are several issues with that:
* `Config` must be instantiated before creating a struct for config params.
* You still need to remember the type of the underlying value so the code becomes really ugly.

The only reason I wanted to do it because it's easy to make mistakes, like typos, that wouldn't be caught at the compilation time (ex. `configKey: "EnableAssetStat"` instead of `configKey: "EnableAssetStats"` - missing `s`).

After thinking about it more, I think we can have a special _setter_ functions for each config value where we set the value but also check the validity so that validation code is in one place. Check [this](https://play.golang.org/p/7FrYFOzFJnL):
* We don't use reflection so typo errors are not possible.
* We also do validation when setting the value.
